project_utils/cache.py

```diff
@@ -215,10 +217,25 @@ def update_cache(self, url_or_name=None, verbose=False):
             except Exception:
                 print(CACHE_FAILED_TO_OPEN.format(dir_name))
                 continue
+
             for remote in repo.remotes:
                 if verbose:
                     print(CACHE_FETCH_REMOTE.format(dir_name, remote.url))
-                remote.fetch(progress=GitProgressHandler())
+                if sha_or_branch is not None and remote.name == dir_name:
+                    print('Fetching ref {}'.format(sha_or_branch))
+                    remote.fetch(refspec=sha_or_branch, progress=GitProgressHandler())
+                else:
+                    remote.fetch(progress=GitProgressHandler())
+
+    def is_sha_cached(self, url_or_name, sha, verbose=False):
+        repo = self._get_repo(self._create_name(url_or_name))
+        output_data = repo.git.execute(['git', 'cat-file', '-e', sha])

```

Comment:
A SHA can be present in multiple branches.
I'd suggest checking if the same branch mentioned in pin file has that SHA.

Reply:
I have implemented this in the cache command itself as cache.py has no knowledge or a pin file.